File: apps/app.music/server/src/main/scala/com/nublic/app/music/server/filewatcher/MusicProcessor.scala
package com.nublic.app.music.server.filewatcher

import java.io.File
import org.apache.commons.io.FilenameUtils
import org.squeryl.PrimitiveTypeMode._
import com.nublic.filewatcher.scala._
import com.nublic.app.music.server.Solr
import com.nublic.app.music.server.model._
import java.io.FileWriter
import java.io.PrintWriter
import java.util.logging.Logger

class MusicProcessor(watcher: FileWatcherActor) extends Processor("music", watcher, true) {
  
  def taggedMimeTypes: List[String] = List(
      // MP4
      "audio/mp4",
      // MP3
      "audio/mpeg", "audio/x-mpeg", "audio/mp3",
      "audio/x-mp3", "audio/mpeg3", "audio/x-mpeg3",
      "audio/mpg", "audio/x-mpg", "audio/x-mpegaudio",
      // OGG
      "audio/ogg", "application/ogg", "audio/x-ogg",
      "application/x-ogg",
      // ASF
      "audio/asf",
      // WMA
      "audio/x-ms-wma",
      // Real
      "audio/rmf", "audio/x-rmf",
      // FLAC
      "audio/flac"
      )
  
  def taggedExtensions: List[String] = List("mp3", "mp4", "ogg", "flac", "wma", "rm")
  
  def supportedMimeTypes: List[String] = taggedMimeTypes ::: List(
      // AAC
      "audio/aac", "audio/x-aac",
      // AC3
      "audio/ac3",
      // AIFF
      "audio/aiff", "audio/x-aiff", "sound/aiff",
      "audio/x-pn-aiff",
      // MIDI
      "audio/mid", "audio/x-midi", 
      // AU
      "audio/basic", "audio/x-basic", "audio/au", 
      "audio/x-au", "audio/x-pn-au", "audio/x-ulaw",
      // PCM
      "application/x-pcm",
      // WAV
      "audio/wav", "audio/x-wav", "audio/wave",
      "audio/x-pn-wav",
      // Various
      "audio/vnd.qcelp", "audio/x-gsm", "audio/snd"
      )
  
  def supportedExtensions: List[String] = taggedExtensions ::: 
    List("wav", "aac", "ac3", "aiff", "mid", "midi", "au", "pcm")
  
  def process(c: FileChange) = {
    Logger.global.severe("Filewatcher: Processing " + c.toString())
    c match {
      // case Created(filename, false)  => process_updated_file(filename)
      case Modified(filename, context, false) => process_updated_file(filename, context)
      case Moved(from, to, context, _)        => process_moved_file(from, to, context)
      case Deleted(filename, _, false)        => process_deleted_file(filename)
      case _                                  => { /* Nothing */ }
    }
  }
  
  def process_updated_file(filename: String, context: String): Unit = {
    val extension = FilenameUtils.getExtension(filename)
    
    Logger.global.severe("Filewatcher: Getting info from " + filename)
    val song_info = 
      if (taggedExtensions.contains(extension) || taggedMimeTypes.contains(Solr.getMimeType(filename))) {
        Some(SongInfo.from(filename, context))
      } else if (supportedExtensions.contains(extension) || supportedMimeTypes.contains(Solr.getMimeType(filename))) {
        Some(FilenameExtractor.from(filename, context))
      } else {
        None
      }
    
    Logger.global.severe("Filewatcher: Adding to database " + filename)
    if (song_info.isDefined) {
      inTransaction {
        Database.songByFilename(filename) match {
          case Some(song) => {
            Logger.global.severe("Filewatcher: Replacing in database " + filename)
            replace_in_database(filename, song.id, song_info.get) 
          }
          case None =>  {
            Logger.global.severe("Filewatcher: Really adding to database " + filename)
            add_to_database(filename, song_info.get)
          }
        }
      }
    }
    
    // Logger.global.severe("Filewatcher: Added to database " + filename)
  }
  
  def process_moved_file(from: String, to: String, context: String): Unit = inTransaction {
    Database.songByFilename(from) match {
      case None       => process_updated_file(to, context)
      case Some(song) => {
        song.file = to
        Database.songs.update(song)
      }
    }
  }
  
  def process_deleted_file(filename: String): Unit = inTransaction {
    Database.songByFilename(filename).map(song => {
      Database.songCollections.deleteWhere(sc => sc.songId === song.id)
      // Remove from playlists
      var sps: List[SongPlaylist] = Database.songPlaylists.where(sp => sp.songId == song.id).toList
      while (!sps.isEmpty) {
        val sp = sps.head
        val pl_pos = sp.position
        Database.songPlaylists.deleteWhere(x => x.songId === sp.songId and x.playlistId === sp.playlistId and x.position === sp.position)
        update(Database.songPlaylists)(x =>
          where(x.playlistId === sp.playlistId and sp.position > pl_pos.~)
          set(x.position := x.position - 1)
        )
        // Try again
        sps = Database.songPlaylists.where(sp => sp.songId == song.id).toList
      }
      Database.songs.deleteWhere(s => s.id === song.id)
    })
  }
  
  // Functions for working with the database
  //
  // Note: Unknown discs or artists are shown with "" in their names
  // ===============================================================
  def ensure_or_create_album(file: String, artistName: Option[String], albumName: Option[String]): Album = {
    val directory = new File(file).getParentFile.getPath
    
    inTransaction {
      // CASE 1. We have a artist name and an album name
      if (artistName != None && albumName != None) {
        // Try to find an album with the same name by the same artist
        find_album_with_artist(artistName.get, albumName.get) match {
          case Some(album) => album // Found an album with same artist and album
          case None        => find_album_by_directory(directory, albumName.get)
        }
      // CASE 2. We have only album name
      } else if (albumName != None /* I have the album name */) { 
        find_album_by_directory(directory, albumName.get)
      // CASE 3. We only have artist name or we have nothing
      } else {
        find_album_with_artist(artistName.getOrElse(""), albumName.getOrElse("")) match {
          case Some(album) => album // Found an album with same artist and album
          case None        => {
            val newAlbum = new Album(albumName.getOrElse(""))
            Database.albums.insert(newAlbum)
            newAlbum
          }
        }
      }
    }
  }
  
  private def find_album_with_artist(artistName: String, albumName: String): Option[Album] = {
    val normalizedArtist = StringUtil.normalize(artistName)
    val normalizedAlbum = StringUtil.normalize(albumName)
    val artistId: Long = Database.artistByNameNormalizing(artistName).map(_.id).getOrElse(-1)
    
    val sameArtistAndAlbum = from(Database.songs, Database.artists, Database.albums)(
      (song, artist, album) =>
        where(song.artistId === artistId and song.albumId === album.id /* join */
              and album.normalized === normalizedAlbum)
        select(album)
      )
    Database.maybe(sameArtistAndAlbum)
  }
  
  private def find_album_by_directory(directory: String, albumName: String): Album = {
    val normalizedAlbum = StringUtil.normalize(albumName)
    // Try to find songs in the same folder with the same album name
    val sameAlbumAndPath = from(Database.songs, Database.albums)(
      (song, album) =>
        where(song.albumId === album.id and album.normalized === normalizedAlbum
              and (song.file like (directory + "/%")))
        select(album)
      )
    Database.maybe(sameAlbumAndPath) match {
      case Some(album) => album
      case None        => {
        // If nothing is found
        val newAlbum = new Album(albumName)
        Database.albums.insert(newAlbum)
        newAlbum
      }
    }
  }
  
  def update_solr(file: String, info: SongInfo) = {
    Solr.getInputDocument(file) match {
      case None      => { /* Cannot update it */ }
      case Some(doc) => {
        doc.setField("title", info.title.getOrElse(""))
        doc.setField("artist", info.artist.getOrElse(""))
        doc.setField("album", info.album.getOrElse(""))
        doc.setField("length", info.length.getOrElse(0))
        info.year match {
          case None    => doc.removeField("year")
          case Some(y) => doc.setField("year", y)
        }
        info.track match {
          case None    => doc.removeField("trackNumber")
          case Some(t) => doc.setField("trackNumber", t)
        }
        Solr.update(doc)
      }
    }
  }

  def add_to_database(file: String, info: SongInfo) = {
    inTransaction {
      Logger.global.severe("Filewatcher: Adding artist " + info.artist.getOrElse(""))
      val artist = Database.ensureInDb(info.artist.getOrElse(""), Database.artists, Database.artistByNameNormalizing, new Artist(_))
      Logger.global.severe("Filewatcher: Adding artist image")
      Images.ensureArtist(artist)
      Logger.global.severe("Filewatcher: Adding album " + info.album.getOrElse(""))
      val album = ensure_or_create_album(file, info.artist, info.album)
      Logger.global.severe("Filewatcher: Adding album image")
      Images.ensureAlbum(new File(file), album, Some(artist))
      Logger.global.severe("Filewatcher: Creating song")
      val song = new Song()
      song.file = file
      song.title = info.title.getOrElse("")
      song.artistId = artist.id
      song.albumId = album.id
      song.length = if (info.length.isDefined) { info.length.get } else { 0 } 
      song.year = info.year
      song.track = info.track
      song.disc_no = info.disc_no
      // Logger.global.severe("Filewatcher: Inserting song")
      try {
        Database.songs.insert(song)
      } catch {
        case e: Throwable => Logger.global.severe("Music Filewatcher exception: " + e.getMessage())
      }
    }
    Logger.global.severe("Filewatcher: Adding to solr")
    update_solr(file, info)
  }
  
  def replace_in_database(file: String, id: Long, info: SongInfo) = {
    inTransaction {
      Database.songs.lookup(id).map(song => {
        val prevArtistId = song.artistId
        val prevAlbumId = song.albumId
        val newArtist = Database.ensureInDb(info.artist.getOrElse(""), Database.artists, Database.artistByNameNormalizing, new Artist(_))
        Images.ensureArtist(newArtist)
        val newAlbum = ensure_or_create_album(song.file, info.artist, info.album)
        Images.ensureAlbum(new File(file), newAlbum, Some(newArtist))
        song.title = info.title.getOrElse("")
        song.artistId = newArtist.id
        song.albumId = newAlbum.id
        song.length = if (info.length.isDefined) { info.length.get } else { 0 } 
        song.year = info.year
        song.track = info.track
        song.disc_no = info.disc_no
        try {
          Database.songs.update(song)
        } catch {
          case e: Throwable => Logger.global.severe("Music Filewatcher exception: " + e.getMessage())
        }
        Database.deleteIfNoAssocInDb(prevArtistId, Database.artists, _.artistId)
        Database.deleteIfNoAssocInDb(prevAlbumId, Database.albums, _.albumId)
      })
    }
    update_solr(file, info)
  }
  
  def remove_from_database(filename: String) = {
    inTransaction {
      Database.songByFilename(filename).map(song => { 
        val artistId = song.artistId
        val albumId = song.albumId
        Database.songs.deleteWhere(s => s.id === song.id)
        Database.deleteIfNoAssocInDb(artistId, Database.artists, _.artistId)
        Database.deleteIfNoAssocInDb(albumId, Database.albums, _.albumId)
      })
    }
  }
}

File: filewatcher/previewer/src/nublic_file_watcher/music_processor.py
import os
import os.path
from unidecode import unidecode
import shutil

#from nublic.filewatcher import FileChange
from nublic.files_and_users import get_file_owner, is_file_shared
from nublic_server.places import (get_mime_type, ensure_cache_folder,
                                  get_cache_folder)
from preview_processor import PreviewProcessor
from nublic_app_music_server.server import app
from nublic_app_music_server.model import (db, Album, Artist,
                                           Song, SongCollection, SongPlaylist)
from song_info import (get_song_info,
                       extract_using_filename)
from nublic_app_music_server.images import (get_artist_folder,
                                            get_album_folder,
                                            ensure_artist_image,
                                            ensure_album_image)

import logging
log = logging.getLogger(__name__)

TAGGED_MIME_TYPES = [
    # MP4
    "audio/mp4",
    # MP3
    "audio/mpeg", "audio/x-mpeg", "audio/mp3",
    "audio/x-mp3", "audio/mpeg3", "audio/x-mpeg3",
    "audio/mpg", "audio/x-mpg", "audio/x-mpegaudio",
    # OGG
    "audio/ogg", "application/ogg", "audio/x-ogg",
    "application/x-ogg",
    # ASF
    "audio/asf",
    # WMA
    "audio/x-ms-wma",
    # Real
    "audio/rmf", "audio/x-rmf",
    # FLAC
    "audio/flac"
]
TAGGED_EXTENSIONS = [".mp3", ".mp4", ".ogg", ".flac", ".wma", ".rm"]

SUPPORTED_MIME_TYPES = TAGGED_MIME_TYPES + [
    # AAC
    "audio/aac", "audio/x-aac",
    # AC3
    "audio/ac3",
    # AIFF
    "audio/aiff", "audio/x-aiff", "sound/aiff",
    "audio/x-pn-aiff",
    # MIDI
    "audio/mid", "audio/x-midi",
    # AU
    "audio/basic", "audio/x-basic", "audio/au",
    "audio/x-au", "audio/x-pn-au", "audio/x-ulaw",
    # PCM
    "application/x-pcm",
    # WAV
    "audio/wav", "audio/x-wav", "audio/wave",
    "audio/x-pn-wav",
    # Various
    "audio/vnd.qcelp", "audio/x-gsm", "audio/snd"
]
SUPPORTED_EXTENSIONS = TAGGED_EXTENSIONS + [".wav", ".aac", ".ac3", ".aiff",
                                            ".mid", ".midi", ".au", ".pcm"]


# Set up processors
class MusicProcessor(PreviewProcessor):
    def __init__(self, logger=None, watcher=''):
        PreviewProcessor.__init__(self)
        log.info('Music processor initialised')
        db.init_app(app)

    def get_id(self):
        return 'music'

    # Inherited methods
    def process_updated(self, filename, is_dir, info=None):
        if not is_dir:
            with app.test_request_context():
                self.process_updated_file(filename)

    def process_deleted(self, filename, is_dir, info=None):
        if not is_dir:
            with app.test_request_context():
                self.process_deleted_file(filename)

    def process_attribs_change(self, filename, is_dir, info=None):
        if not is_dir:
            with app.test_request_context():
                filename_unicode = unicode(filename, 'utf8')
                song = Song.query.filter_by(file=filename_unicode).first()
                if song is not None:
                    self.update_attribs(song)
                else:
                    self.process_updated_file(filename)

    def process_moved(self, filename_from, filename_to, is_dir, info=None):
        with app.test_request_context():
            if is_dir:
                self.process_moved_dir(filename_from, filename_to)
            else:
                self.process_moved_file(filename_from, filename_to)

    def is_hidden(self, path):
        """
        path is an byte string in utf8
        """
        filename = os.path.basename(path)
        return filename.endswith('~') or filename.startswith('.')

    def create_view(self, filename):
        log.warning("No compatibility check for audio file %s", filename)
        cache_folder = ensure_cache_folder(filename)  # @TODO: Check for utf8
        mp3_cache = os.path.join(cache_folder, 'view.mp3')
        if os.path.exists(mp3_cache):
            # @TODO Check if it is a symlink and if it is not if the
            # view is newer than the cache
            log.warning("Cache existing for audio file %s, skipping", filename)
        else:
            log.info("Creating cache for audio file %s", filename)
            # @TODO Do conversion if needed
            os.symlink(filename, mp3_cache)

    def process_updated_file(self, filename):
        ''' Process an update of a file.
        The file might exist before.
        filename is an byte string in utf8
        '''
        if os.path.exists(filename):
            filename_unicode = unicode(filename, 'utf8')
            log.info('Updated file: %s', filename)
            mime = get_mime_type(filename)
            #_, ext = os.path.splitext(filename)
            # Process song
            if mime in TAGGED_MIME_TYPES:  # or ext in TAGGED_EXTENSIONS:
                song_info = get_song_info(filename)
            elif mime in SUPPORTED_MIME_TYPES:  # or ext in SUPPORTED_EXTENSIONS:
                song_info = extract_using_filename(
                    filename).clean()
            else:
                song_info = None
            # If we got some information
            if song_info is not None:
                self.create_view(filename)
                # Add to database now
                s = Song.query.filter_by(file=filename_unicode).first()
                if s is None:
                    self.add_to_database(filename, song_info)
                else:
                    self.replace_in_database(filename, s, song_info)

    def update_attribs(self, song):
        """
        Updates the attributes in the Song object in the database
        song.file is an Unicode object
        """
        song.owner = get_file_owner(song.file.encode('utf-8')).get_username()
        song.shared = is_file_shared(song.file.encode('utf-8'))
        db.session.commit()

    def delete_song_from_collections(self, song_id):
        """ Delete from collections """
        SongCollection.query.filter_by(songId=song_id).delete()
        db.session.commit()

    def delete_song_from_playlists(self, song_id):
        " Delete from playlists "
        relation = SongPlaylist.query.filter_by(songId=song_id).first()
        while relation is not None:
            rest = SongPlaylist.query.filter_by(playlistId=relation.playlistId).filter(SongPlaylist.position > relation.position).all()
            for other_song in rest:
                other_song.position -= 1
            db.session.delete(relation)
            #relation.delete()
            db.session.commit()
            # Try again
            relation = SongPlaylist.query.filter_by(songId=song_id).first()

    def remove_view(self, filename):
        # exists return false for broken symlinks so we need to check is link
        # before exists to remove the broken symlink
        mp3_cache = os.path.join(get_cache_folder(filename), 'view.mp3')
        if os.path.islink(mp3_cache) or os.path.exists(mp3_cache):
            log.info(
                "Cache removed for audio file %s as %s", filename, mp3_cache)
            os.remove(mp3_cache)
        else:
            log.warning("Cache for audio file %s does not exist in %s. Not removed",
                        filename, mp3_cache)

    def process_deleted_file(self, filename):
        """
        Process a file deleted by removing the file from the previews
        and the database
        filename is a byte string in utf8
        """
        filename_unicode = unicode(filename, 'utf8')
        song = Song.query.filter_by(file=filename_unicode).first()
        if song is not None:
            self.delete_song_from_collections(song.id)
            self.delete_song_from_playlists(song.id)
            # Delete the song itself
            artistId = song.artistId
            albumId = song.albumId
            db.session.delete(song)
            db.session.commit()
            self.delete_artist_if_no_assoc(artistId)
            self.delete_album_if_no_assoc(albumId)
            self.remove_view(filename)

    def process_moved_dir(self, from_, to):
        """
        from_ and to are byte strings in utf8
        """
        if os.path.exists(to):  # The dir was removed before this. Nothing to do
            for e in os.listdir(to):
                file_from = os.path.join(from_, e)
                file_to = os.path.join(to, e)
                if os.path.isdir(file_to) and not self.is_hidden(file_to):
                    self.process_moved_dir(file_from, file_to)
                else:
                    self.process_moved_file(file_from, file_to)

    def process_moved_file(self, from_, to):
        """
        from_ and to are byte strings in utf8
        """
        from_unicode = unicode(from_, 'utf8')
        song = Song.query.filter_by(file=from_unicode).first()
        if song is not None:
            song.file = unicode(to, 'utf8')
            db.session.commit()
            # @TODO Do not recreate view if not needed
            self.remove_view(from_)
            self.create_view(to)
        else:
            self.process_updated_file(to)

    def ensure_or_create_artist(self, artist_name):
        """
        Returns an artist.
        If the artist does not exists it is created.
        This function tries to match close but not exact artists
        """
        n_artist = unidecode(unicode(artist_name).lower())
        # Try to find an artist
        artist = Artist.query.filter_by(normalized=n_artist).first()
        if artist is not None:
            return artist
        else:
            new_artist = Artist(artist_name)
            db.session.add(new_artist)
            db.session.commit()
            return new_artist

    def ensure_or_create_album(self, file_, artist_name, album_name):
        """
        Returns an album for the artist given.
        If the album does not exist the album is created for the
        given file_ before.
        file_ is a byte string in utf8
        """
        directory, _ = os.path.split(file_)
        # Case 1. we have artist and album name
        if artist_name is not None and album_name is not None:
            ab = self.find_album_with_artist(artist_name, album_name)
            if ab is not None:
                return ab
            else:
                return self.find_album_by_directory(directory, album_name)
        elif album_name is not None:
            return self.find_album_by_directory(directory, album_name)
        else:
            if artist_name is None:
                r_artist_name = ''
            else:
                r_artist_name = artist_name
            if album_name is None:
                r_album_name = ''
            else:
                r_album_name = artist_name
            ab = self.find_album_with_artist(r_artist_name, r_album_name)
            if ab is not None:
                return ab
            else:
                new_album = Album(r_album_name)
                db.session.add(new_album)
                db.session.commit()
                return new_album

    def find_album_with_artist(self, artist_name, album_name):
        n_artist = unidecode(unicode(artist_name).lower())
        n_album = unidecode(unicode(album_name).lower())
        # Try to get artist
        artist = Artist.query.filter_by(normalized=n_artist).first()
        if artist is None:
            return None
        else:
            return Album.query.filter_by(normalized=n_album).filter(Song.albumId == Album.id).filter(Song.artistId == artist.id).first()

    def find_album_by_directory(self, directory, album_name):
        n_album = unidecode(unicode(album_name).lower())
        # Try to find a song with name album name
        ab = Album.query.filter_by(normalized=n_album).filter(Song.albumId == Album.id).filter(Song.file.like(directory + '/%')).first()
        if ab is not None:
            return ab
        else:
            new_album = Album(album_name)
            db.session.add(new_album)
            db.session.commit()
            return new_album

    def delete_artist_if_no_assoc(self, artist_id):
        songs = Song.query.filter_by(artistId=artist_id).count()
        if songs == 0:
            Artist.query.filter_by(id=artist_id).delete()
            db.session.commit()
            shutil.rmtree(get_artist_folder(artist_id))

    def delete_album_if_no_assoc(self, album_id):
        songs = Song.query.filter_by(albumId=album_id).count()
        if songs == 0:
            Album.query.filter_by(id=album_id).delete()
            db.session.commit()
            shutil.rmtree(get_album_folder(album_id))

    def add_to_database(self, filename, song_info):
        ''' Add the song to the database if required.
        filename is a byte string in utf8
        '''
        filename_unicode = unicode(filename, 'utf8')
        # Just in case anything is None
        if song_info.title is None:
            r_title = ''
        else:
            r_title = song_info.title
        if song_info.artist is None:
            r_artist_name = ''
        else:
            r_artist_name = song_info.artist
        if song_info.album is None:
            r_album_name = ''
        else:
            r_album_name = song_info.album
        if song_info.length is None:
            r_length = 0
        else:
            r_length = song_info.length
        log.info('Adding to database: %s by %s in %s', r_title,
                 r_artist_name, r_album_name)
        # Ensure artist
        artist = self.ensure_or_create_artist(r_artist_name)
        ensure_artist_image(artist.id, r_artist_name)
        # Ensure album
        album = self.ensure_or_create_album(
            filename, r_artist_name, r_album_name)
        ensure_album_image(album.id, filename, r_album_name, r_artist_name)
        # Create song in database
        song = Song(filename_unicode, r_title, artist.id, album.id, r_length,
                    song_info.year, song_info.track, song_info.disc_no)
        db.session.add(song)
        db.session.commit()

    def replace_in_database(self, filename, s, song_info):
        """
        Replace a song that changed in the database
        filename is a byte string in utf8
        """
        # Just in case anything is None
        if song_info.title is None:
            r_title = ''
        else:
            r_title = song_info.title
        if song_info.artist is None:
            r_artist_name = ''
        else:
            r_artist_name = song_info.artist
        if song_info.album is None:
            r_album_name = ''
        else:
            r_album_name = song_info.album
        if song_info.length is None:
            r_length = 0
        else:
            r_length = song_info.length
        # Save previous artist and album
        prev_artist_id = s.artistId
        prev_album_id = s.albumId
        # Ensure artist
        artist = self.ensure_or_create_artist(r_artist_name)
        ensure_artist_image(artist.id, r_artist_name)
        # Ensure album
        album = self.ensure_or_create_album(
            filename, r_artist_name, r_album_name)
        ensure_album_image(album.id, filename, r_album_name, r_artist_name)
        # Change song object
        s.title = r_title
        s.artistId = artist.id
        s.albumId = album.id
        s.length = r_length
        s.year = song_info.year
        s.track = song_info.track
        s.disc_no = song_info.disc_no
        db.session.commit()
        # Delete previous artist and album if not needed
        self.delete_artist_if_no_assoc(prev_artist_id)
        self.delete_album_if_no_assoc(prev_album_id)
